### PR TITLE
Move CloudTrail trail to default region (#4831)

### LINKED ISSUE #4831: Move CloudTrail trail to default region
We're constantly running into issues with our choice to place the trail into the US-west-2 region (which we did to maintain continuity with the preexisting trail). Other than continuity, we gain nothing from that choice. 

- [x] Create a new trail called `azul-cloudtrail-{deployment_stage}`<sup>1</sup>. Leave the original trail in place.
- [x] Change all references to the original trail to point to the new trail.
- [x] Disable logging on the original trail
- [x] Create a new trail bucket with the same name<sup>1</sup> as the current bucket with the suffix `.{region}` appended. 
- [x] Associate the new trail with the new bucket. 

<sup>1 Name changed from `cloudtrail` to `trail` to ensure consistency and avoid conflict with existing log group</sup>

-----

- [ ] Security design review completed; the Resolution of this issue does **not** …
  - [ ] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with 
  - [ ] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [ ] … affect the permissions of external users like access to
    - TDR snapshots
  - [ ] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [ ] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [ ] … affect monitoring of the system
  - [ ] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [ ] … add an interface that exposes sensitive or confidential data at the security boundary
  - [ ] … affect the encryption of data at rest
  - [ ] … require persistence of sensitive or confidential data that might require encryption at rest
  - [ ] … require unencrypted transmission of data within the security boundary
  - [ ] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [ ] Documentation on any unchecked boxes is provided in comments below

### PR DESCRIPTION
**Pull request by nadove-ucsc**
Connected issues: #4831 

### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references all connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] For each connected issue, there is at least one commit whose title references that issue
- [x] PR is connected to all connected issues via ZenHub
- [x] PR description links to connected issues
- [x] Added `partial` label to PR <sub>or this PR completely resolves all connected issues</sub>

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title

### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR <sub>or this PR does not require reindexing</sub>
- [x] Added `a` (compatible changes) or `A` (incompatible ones) tag to commit title <sub>or this PR does not modify the Azul service API</sub>
- [x] Added `API` label to connected issues <sub>or this PR does not modify the Azul service API</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the `prod` branch has no temporary hotfixes for any connected issues</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR <sub>or this PR does not touch requirements*.txt</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>

### Peer reviewer (after requesting changes)

Uncheck the *Author (before every review)* checklists.

### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *before every review* checklists. Update the `N reviews` label.

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Pushed PR branch to GitLab `dev` and added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved connected issues to Merged column in ZenHub
- [x] Pushed merge commit to GitHub

### Operator (after pushing the merge commit)

- [x] Shortened the PR chain <sub>or this PR is not labeled `base`</sub>
- [x] Followed upgrading instructions in `UPGRADING.rst` for `dev`
- [x] Pushed merge commit to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Followed upgrading instructions in `UPGRADING.rst` for `anvildev`
- [x] Pushed merge commit to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes on GitLab `dev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `dev`<sup>1</sup>
- [x] Build passes on GitLab `anvildev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvildev`<sup>1</sup>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is*
pushed determines this checklist item.

### Operator (reindex)

- [x] Deleted unreferenced indices in `dev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvildev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `dev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvildev` deployment <sub>or this PR does not require reindexing</sub>

### Operator

- [x] Added item to PR checklist for prod promotion: "Followed upgrading instructions in `UPGRADING.rst` for `prod`"
- [x] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Comment by nadove-ucsc** [comment]
This PR:
- Creates the new log group
- Enables logging on the new bucket and trail
- Disables logging on the old bucket and trail

File: terraform/shared/shared.tf.json.template.py [hunk 1]
```diff
@@ -181,7 +194,14 @@
                 's3_bucket_name': '${aws_s3_bucket.shared_cloudtrail.id}',
                 'enable_log_file_validation': True,
                 'is_multi_region_trail': True,
```
Comment (achave11-ucsc) [inline terraform/shared/shared.tf.json.template.py]: Consider adding …
```suggestion
                'is_multi_region_trail': True,
                'enable_logging': False,
```
… [tf cloudtrail enable_logging](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/cloudtrail#enable_logging) to explicitly pause logging, #4831 prescribes disabling so I am unsure if they are synonymous.

**Review by achave11-ucsc — COMMENTED** [review]
Changes look good. However, checklist items in https://github.com/DataBiosphere/azul/issues/4831#issue-1497346946 remain unchecked, consider updating them. It seems to me like they have been completed.

**Review by achave11-ucsc — APPROVED** [review]

File: terraform/shared/shared.tf.json.template.py [hunk 2]
```diff
@@ -50,36 +58,42 @@
...
-                            'Resource': '${aws_s3_bucket.shared_cloudtrail.arn}'
-                        },
-                        {
-                            'Effect': 'Allow',
-                            'Principal': {
-                                'Service': 'cloudtrail.amazonaws.com'
+            **{
+                bucket: {
+                    **provider_fragment(region),
+                    'bucket': f'${{aws_s3_bucket.{bucket}.id}}',
```
Comment (hannes-ucsc) [inline terraform/shared/shared.tf.json.template.py]: Precedent (at least in this file) is to use `%` operator when the string already contains braces from TF interpolation.

File: terraform/shared/shared.tf.json.template.py [hunk 3]
```diff
@@ -175,21 +189,36 @@
             }
         },
         'aws_cloudtrail': {
+            # FIXME: Disable original CloudTrail trail #4832
```
Comment (hannes-ucsc) [inline terraform/shared/shared.tf.json.template.py]:
```suggestion
            # FIXME: Disable original CloudTrail trail
```

File: terraform/shared/shared.tf.json.template.py [hunk 4]
```diff
@@ -175,21 +189,36 @@
...
+            #        https://github.com/databiosphere/azul/issues/4832
             'shared': {
                 **provider_fragment(config.cloudtrail_trail_region),
                 'name': 'azul-shared',
                 's3_bucket_name': '${aws_s3_bucket.shared_cloudtrail.id}',
```
Comment (hannes-ucsc) [inline terraform/shared/shared.tf.json.template.py:196]:
```suggestion
                's3_bucket_name': '${aws_s3_bucket.shared_cloudtrail.id}',
                'enable_logging': False,
```

File: terraform/shared/shared.tf.json.template.py [hunk 5]
```diff
@@ -175,21 +189,36 @@
...
-                'cloud_watch_logs_group_arn': '${aws_cloudwatch_log_group.cloudtrail.arn}:*',
+                'enable_logging': False,
+            },
+            'trail': {
+                'name': config.qualified_resource_name('trail'),
+                's3_bucket_name': '${aws_s3_bucket.trail.id}',
+                'enable_log_file_validation': True,
+                'is_multi_region_trail': True,
+                'cloud_watch_logs_group_arn': '${aws_cloudwatch_log_group.trail.arn}:*',
                 'cloud_watch_logs_role_arn': '${aws_iam_role.cloudtrail.arn}'
```
Comment (hannes-ucsc) [inline terraform/shared/shared.tf.json.template.py]:
```suggestion
                'cloud_watch_logs_role_arn': '${aws_iam_role.trail.arn}'
```



Ideally, all CloudTrail-related resources should have the `trail` TF resource name and a fully-qualified real-world name like `azul-trail-dev`. Except the trail bucket, that is.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**